File: src-tauri/src/whatsapp.rs
use serde::Serialize;
use std::io::Write;
use std::sync::Mutex;
use tauri::{AppHandle, Manager, State};

use crate::{
    auth::require_admin, common::get_db, common::DbConn, common::record_app_log, session::SessionState,
};
use rusqlite::params;

#[derive(Debug, Serialize)]
pub struct WhatsAppSidecarStatus {
    pub status: String,
    pub qr_data_url: Option<String>,
    pub last_error: Option<String>,
    pub has_client: bool,
    pub enabled: bool,
    pub auto_notify_owner: bool,
    pub owner_number: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct WaNotifyResponse {
    pub sent: bool,
    pub reason: Option<String>,
    pub id: Option<String>,
}

pub struct WaSidecarState {
    pub child: Mutex<Option<std::process::Child>>,
    pub status: Mutex<String>,
    pub qr_data_url: Mutex<Option<String>>,
    pub last_error: Mutex<Option<String>>,
}

impl WaSidecarState {
    pub fn new() -> Self {
        Self {
            child: Mutex::new(None),
            status: Mutex::new("idle".into()),
            qr_data_url: Mutex::new(None),
            last_error: Mutex::new(None),
        }
    }
}

fn get_whatsapp_settings(conn: &rusqlite::Connection) -> (bool, bool, String) {
    let mut stmt = conn.prepare("SELECT key, value FROM settings").ok();
    let mut enabled = false;
    let mut auto_notify = false;
    let mut owner_number = String::new();
    if let Some(ref mut s) = stmt {
        let rows = s.query_map([], |row| {
            Ok((row.get::<_, String>(0)?, row.get::<_, String>(1)?))
        });
        if let Ok(rows) = rows {
            for row in rows.flatten() {
                match row.0.as_str() {
                    "whatsapp_enabled" => enabled = row.1 == "true",
                    "whatsapp_auto_notify_owner" => auto_notify = row.1 == "true",
                    "whatsapp_owner_number" => owner_number = row.1,
                    _ => {}
                }
            }
        }
    }
    (enabled, auto_notify, owner_number)
}

/// Kill the existing sidecar process if running
fn kill_sidecar(sc: &WaSidecarState) {
    if let Ok(mut child) = sc.child.lock() {
        if let Some(ref mut c) = *child {
            let _ = c.kill();
            let _ = c.wait();
        }
        *child = None;
    }
}

/// Send a JSONL command to the sidecar's stdin
fn send_sidecar_command(
    sc: &WaSidecarState,
    command: &str,
    payload: &serde_json::Value,
) -> Result<(), String> {
    let mut child_guard = sc.child.lock().map_err(|_| "lock error".to_string())?;
    let child = child_guard
        .as_mut()
        .ok_or("Sidecar tidak berjalan".to_string())?;
    let stdin = child
        .stdin
        .as_mut()
        .ok_or("Stdin sidecar tidak tersedia".to_string())?;

    let mut msg = serde_json::Map::new();
    msg.insert("cmd".into(), serde_json::Value::String(command.into()));
    if let Some(obj) = payload.as_object() {
        for (k, v) in obj {
            msg.insert(k.clone(), v.clone());
        }
    }
    let line = serde_json::to_string(&msg).map_err(|e| format!("Gagal serialisasi: {e}"))?;
    writeln!(stdin, "{}", line).map_err(|e| format!("Gagal kirim ke sidecar: {e}"))?;
    stdin
        .flush()
        .map_err(|e| format!("Gagal flush sidecar: {e}"))?;
    Ok(())
}

#[tauri::command]
pub fn whatsapp_status(
    app: AppHandle,
    session: State<'_, SessionState>,
    db: State<'_, DbConn>,
) -> Result<WhatsAppSidecarStatus, String> {
    let _user = require_admin(&session)?;
    let sc = app.state::<WaSidecarState>();
    let conn = get_db(&db)?;
    let (enabled, auto_notify, owner_number) = get_whatsapp_settings(&conn);
    let status = sc.status.lock().map(|s| s.clone()).unwrap_or_default();
    let qr = sc.qr_data_url.lock().ok().and_then(|q| q.clone());
    let err = sc.last_error.lock().ok().and_then(|e| e.clone());
    let has_client = sc.child.lock().ok().map(|c| c.is_some()).unwrap_or(false);
    Ok(WhatsAppSidecarStatus {
        status,
        qr_data_url: qr,
        last_error: err,
        has_client,
        enabled,
        auto_notify_owner: auto_notify,
        owner_number: if owner_number.is_empty() {
            None
        } else {
            Some(owner_number)
        },
    })
}

#[tauri::command]
pub fn whatsapp_start(
    app: AppHandle,
    session: State<'_, SessionState>,
    db: State<'_, DbConn>,
) -> Result<WhatsAppSidecarStatus, String> {
    let _user = require_admin(&session)?;
    let sc = app.state::<WaSidecarState>();

    // Kill existing sidecar if any
    kill_sidecar(&sc);
    *sc.status.lock().map_err(|_| "lock error".to_string())? = "initializing".into();
    *sc.qr_data_url
        .lock()
        .map_err(|_| "lock error".to_string())? = None;
    *sc.last_error.lock().map_err(|_| "lock error".to_string())? = None;

    // Determine wa-service path: try bundled sidecar first, then local dev
    let child = match app
        .path()
        .resource_dir()
        .ok()
        .map(|dir| dir.join("wa-service"))
        .filter(|p| p.exists())
    {
        Some(path) => {
            // Bundled sidecar binary
            std::process::Command::new(path)
                .stdin(std::process::Stdio::piped())
                .stdout(std::process::Stdio::piped())
                .stderr(std::process::Stdio::piped())
                .spawn()
                .map_err(|e| format!("Gagal spawn sidecar: {e}"))?
        }
        None => {
            // Fallback: run via node from local wa-service/ directory
            let project_root = std::env::current_dir()
                .ok()
                .and_then(|d| d.parent().map(|p| p.to_path_buf()))
                .unwrap_or_default();
            let wa_dir = project_root.join("wa-service");
            let wa_entry = wa_dir.join("index.mjs");

            if !wa_entry.exists() {
                *sc.status.lock().map_err(|_| "lock error".to_string())? = "error".into();
                *sc.last_error.lock().map_err(|_| "lock error".to_string())? = Some(
                    "wa-service tidak ditemukan. Pastikan folder wa-service/index.mjs ada.".into(),
                );
                return whatsapp_status(app, session, db);
            }

            std::process::Command::new("node")
                .arg(&wa_entry)
                .stdin(std::process::Stdio::piped())
                .stdout(std::process::Stdio::piped())
                .stderr(std::process::Stdio::piped())
                .spawn()
                .map_err(|e| format!("Gagal spawn wa-service: {e}"))?
        }
    };

    *sc.child.lock().map_err(|_| "lock error".to_string())? = Some(child);
    let conn = get_db(&db)?;
    record_app_log(
        &conn,
        "INFO",
        "whatsapp",
        "WhatsApp sidecar dimulai",
    );

    whatsapp_status(app, session, db)
}

#[tauri::command]
pub fn whatsapp_restart(
    app: AppHandle,
    session: State<'_, SessionState>,
    db: State<'_, DbConn>,
) -> Result<WhatsAppSidecarStatus, String> {
    let _user = require_admin(&session)?;
    let sc = app.state::<WaSidecarState>();
    kill_sidecar(&sc);
    *sc.status.lock().map_err(|_| "lock error".to_string())? = "idle".into();
    *sc.qr_data_url
        .lock()
        .map_err(|_| "lock error".to_string())? = None;
    *sc.last_error.lock().map_err(|_| "lock error".to_string())? = None;
    let conn = get_db(&db)?;
    record_app_log(
        &conn,
        "INFO",
        "whatsapp",
        "WhatsApp sidecar di-restart",
    );
    whatsapp_status(app, session, db)
}

#[tauri::command]
pub fn whatsapp_logout(app: AppHandle, session: State<'_, SessionState>, db: State<'_, DbConn>) -> Result<bool, String> {
    let _user = require_admin(&session)?;
    let sc = app.state::<WaSidecarState>();

    // Send logout command before killing
    let _ = send_sidecar_command(&sc, "logout", &serde_json::json!({}));

    kill_sidecar(&sc);
    *sc.status.lock().map_err(|_| "lock error".to_string())? = "disconnected".into();
    *sc.qr_data_url
        .lock()
        .map_err(|_| "lock error".to_string())? = None;
    *sc.last_error.lock().map_err(|_| "lock error".to_string())? = None;
    let conn = get_db(&db)?;
    record_app_log(&conn, "INFO", "whatsapp", "WhatsApp logout");
    Ok(true)
}

#[tauri::command]
pub fn whatsapp_notify(
    app: AppHandle,
    session: State<'_, SessionState>,
    db: State<'_, DbConn>,
    transaction_id: i64,
) -> Result<WaNotifyResponse, String> {
    let _user = require_admin(&session)?;
    let sc = app.state::<WaSidecarState>();
    let conn = get_db(&db)?;
    let (enabled, auto_notify, owner_number) = get_whatsapp_settings(&conn);

    if !enabled || !auto_notify {
        return Ok(WaNotifyResponse {
            sent: false,
            reason: Some("disabled".into()),
            id: None,
        });
    }
    if owner_number.is_empty() {
        return Ok(WaNotifyResponse {
            sent: false,
            reason: Some("missing_owner_number".into()),
            id: None,
        });
    }

    // Build notification message based on transaction type
    let trx = conn
        .query_row(
            "SELECT invoice_no, type, total_amount, customer_name, status, created_at FROM transactions WHERE id = ?1",
            params![transaction_id],
            |row| {
                Ok((
                    row.get::<_, String>(0)?,
                    row.get::<_, String>(1)?,
                    row.get::<_, f64>(2)?,
                    row.get::<_, Option<String>>(3)?,
                    row.get::<_, String>(4)?,
                    row.get::<_, String>(5)?,
                ))
            },
        )
        .map_err(|_| "Transaksi tidak ditemukan".to_string())?;

    let message = build_notification_message(&trx.1, &trx.0, trx.2, &trx.3, &trx.4, &trx.5);
    let to = format!("{}@s.whatsapp.net", owner_number);

    // Check if sidecar is running and connected
    let has_child = sc.child.lock().ok().map(|c| c.is_some()).unwrap_or(false);
    let status = sc.status.lock().map(|s| s.clone()).unwrap_or_default();

    if !has_child || status != "connected" {
        return Ok(WaNotifyResponse {
            sent: false,
            reason: Some(format!(
                "sidecar_{}",
                if !has_child {
                    "not_running"
                } else {
                    "not_connected"
                }
            )),
            id: None,
        });
    }

    // Send via sidecar IPC
    let payload = serde_json::json!({
        "to": to,
        "message": message,
    });
    send_sidecar_command(&sc, "send", &payload)?;

    record_app_log(
        &conn,
        "INFO",
        "whatsapp",
        &format!("WA notif dikirim untuk trx #{}", transaction_id),
    );
    Ok(WaNotifyResponse {
        sent: true,
        reason: None,
        id: None,
    })
}

fn build_notification_message(
    trx_type: &str,
    invoice_no: &str,
    amount: f64,
    customer: &Option<String>,
    status: &str,
    created_at: &str,
) -> String {
    let customer_str = customer.as_deref().unwrap_or("-");
    match trx_type {
        "pos" => format!(
            "[NOTIFIKASI TRANSAKSI POS]\n\nInvoice: {}\nTipe: Penjualan POS\nNominal: Rp{:.0}\nPelanggan: {}\nStatus: {}\nTanggal: {}\n\nTercatat otomatis oleh CatatAgen Local",
            invoice_no, amount, customer_str, status, created_at
        ),
        "brilink" => format!(
            "[NOTIFIKASI TRANSAKSI BRILINK]\n\nInvoice: {}\nTipe: Layanan Agen\nNominal: Rp{:.0}\nPelanggan: {}\nStatus: {}\nTanggal: {}\n\nTercatat otomatis oleh CatatAgen Local",
            invoice_no, amount, customer_str, status, created_at
        ),
        _ => format!(
            "[NOTIFIKASI TRANSAKSI]\n\nInvoice: {}\nTipe: {}\nNominal: Rp{:.0}\nPelanggan: {}\nStatus: {}\nTanggal: {}\n\nTercatat otomatis oleh CatatAgen Local",
            invoice_no, trx_type, amount, customer_str, status, created_at
        ),
    }
}

File: src/routes/whatsapp.js
/**
 * WhatsApp routes — integrated Baileys (no separate sidecar process).
 *
 * Baileys runs in-process, managed via these API endpoints.
 * Session files stored in app data directory.
 */

import { Router } from 'express';
import QRCode from 'qrcode';
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { getAppDir, getDb, logError, recordLog } from '../config/database.js';
import { requireAuth, requireAdmin } from '../middleware/auth.js';

const router = Router();

// ── State ──────────────────────────────────────────────────────────
let sock = null;
let connectionStatus = 'idle';
let qrDataUrl = null;
let lastError = null;
let shouldReconnect = true;
let startTime = Date.now();

const SESSION_DIR = path.join(getAppDir(), 'whatsapp-session');

function setStatus(s, err) {
  connectionStatus = s;
  if (err !== undefined) lastError = err || null;
  console.log(`[WA] Status: ${s}${err ? ' — ' + err : ''}`);
}

function normalizePhone(phone) {
  let digits = String(phone || '').replace(/\D/g, '');
  if (!digits) return '';
  if (digits.startsWith('0')) return '62' + digits.slice(1);
  if (digits.startsWith('62')) return digits;
  if (digits.startsWith('8')) return '62' + digits;
  return digits;
}

// ── WhatsApp Connection ──────────────────────────────────────────

async function connectToWhatsApp() {
  if (sock || connectionStatus === 'initializing' || connectionStatus === 'connecting') return;

  setStatus('initializing');
  lastError = null;
  qrDataUrl = null;

  try {
    fs.mkdirSync(SESSION_DIR, { recursive: true, mode: 0o700 });
  } catch {}

  try {
    const { makeWASocket, useMultiFileAuthState, DisconnectReason, Browsers } = await import('@whiskeysockets/baileys');
    const { state, saveCreds } = await useMultiFileAuthState(SESSION_DIR);

    sock = makeWASocket({
      auth: state,
      browser: Browsers.ubuntu('Chrome'),
      printQRInTerminal: false,
      logger: {
        level: 'silent',
        info: () => {}, debug: () => {},
        warn: (msg) => console.warn('[WA]', msg),
        error: (msg) => console.error('[WA]', msg),
        trace: () => {}, fatal: (msg) => console.error('[WA] FATAL:', msg),
        child: () => ({
          level: 'silent', info: () => {}, debug: () => {},
          warn: () => {}, error: () => {}, trace: () => {},
          fatal: () => {}, child: () => {},
        }),
      },
      shouldSyncHistoryMessage: () => false,
      getMessage: () => Promise.resolve({}),
    });

    sock.ev.on('connection.update', async (update) => {
      const { connection, lastDisconnect, qr } = update;

      if (qr) {
        console.log('[WA] QR received');
        try {
          qrDataUrl = await QRCode.toDataURL(qr, { margin: 1, width: 320 });
          setStatus('qr');
        } catch (e) {
          console.error('[WA] QR encode failed:', e.message);
        }
      }

      if (connection === 'close') {
        const code = lastDisconnect?.error?.output?.statusCode;
        const msg = lastDisconnect?.error?.message || 'Connection closed';
        console.log(`[WA] Connection closed: ${code} — ${msg}`);
        sock = null;

        if (code === DisconnectReason.loggedOut) {
          setStatus('disconnected', 'Session di-logout dari perangkat lain');
        } else if (shouldReconnect) {
          setStatus('connecting', msg);
          console.log('[WA] Reconnecting in 3s...');
          setTimeout(() => connectToWhatsApp(), 3000);
        } else {
          setStatus('disconnected', msg);
        }
      }

      if (connection === 'open') {
        qrDataUrl = null;
        setStatus('ready');
        console.log('[WA] Connected & ready!');
      }
    });

    sock.ev.on('creds.update', saveCreds);
    setStatus('connecting');
  } catch (err) {
    console.error('[WA] Init failed:', err.message);
    setStatus('error', err.message);
    sock = null;
  }
}

async function doSend(to, message) {
  const number = normalizePhone(to);
  if (!number) throw new Error('Nomor WhatsApp tujuan belum diatur');
  if (!sock || connectionStatus !== 'ready') {
    throw new Error(`WhatsApp belum siap (status: ${connectionStatus})`);
  }
  const jid = `${number}@s.whatsapp.net`;
  await sock.sendMessage(jid, { text: message });
}

async function doLogout() {
  if (sock) {
    try { await sock.logout(); } catch {}
    try { sock.end(); } catch {}
  }
  sock = null;
  qrDataUrl = null;
  setStatus('disconnected');
  try {
    const files = fs.readdirSync(SESSION_DIR);
    for (const f of files) {
      fs.rmSync(path.join(SESSION_DIR, f), { recursive: true, force: true });
    }
  } catch {}
}

async function doRestart() {
  if (sock) {
    try { sock.end(); } catch {}
    sock = null;
  }
  qrDataUrl = null;
  lastError = null;
  connectionStatus = 'idle';
  await connectToWhatsApp();
}

function getWhatsAppSettings() {
  const db = getDb();
  const rows = db.prepare('SELECT key, value FROM settings').all();
  let enabled = false;
  let autoNotify = false;
  let ownerNumber = '';
  for (const row of rows) {
    if (row.key === 'whatsapp_enabled') enabled = row.value === 'true';
    if (row.key === 'whatsapp_auto_notify_owner') autoNotify = row.value === 'true';
    if (row.key === 'whatsapp_owner_number') ownerNumber = row.value;
  }
  return { enabled, autoNotify, ownerNumber };
}

// ── Routes ──────────────────────────────────────────────────────────

// GET /api/whatsapp/status
router.get('/whatsapp/status', requireAuth, requireAdmin, (_req, res) => {
  try {
    const { enabled, autoNotify, ownerNumber } = getWhatsAppSettings();

    res.json({
      status: connectionStatus,
      qr_data_url: connectionStatus === 'qr' ? qrDataUrl : null,
      last_error: lastError,
      has_client: !!sock,
      enabled,
      auto_notify_owner: autoNotify,
      owner_number: ownerNumber || null,
    });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
});

// POST /api/whatsapp/start
router.post('/whatsapp/start', requireAuth, requireAdmin, async (_req, res) => {
  try {
    await connectToWhatsApp();
    recordLog(getDb(), 'INFO', 'whatsapp', 'WhatsApp dimulai');

    // Return current status after brief wait
    setTimeout(() => {
      res.json({
        status: connectionStatus,
        qr_data_url: connectionStatus === 'qr' ? qrDataUrl : null,
        last_error: lastError,
        has_client: !!sock,
      });
    }, 500);
  } catch (err) {
    logError(getDb(), 'whatsapp', `Start error: ${err.message}`);
    res.status(500).json({ error: err.message });
  }
});

// POST /api/whatsapp/restart
router.post('/whatsapp/restart', requireAuth, requireAdmin, async (_req, res) => {
  try {
    await doRestart();
    recordLog(getDb(), 'INFO', 'whatsapp', 'WhatsApp di-restart');

    setTimeout(() => {
      res.json({
        status: connectionStatus,
        qr_data_url: connectionStatus === 'qr' ? qrDataUrl : null,
        last_error: lastError,
        has_client: !!sock,
      });
    }, 500);
  } catch (err) {
    logError(getDb(), 'whatsapp', `Restart error: ${err.message}`);
    res.status(500).json({ error: err.message });
  }
});

// POST /api/whatsapp/logout
router.post('/whatsapp/logout', requireAuth, requireAdmin, async (_req, res) => {
  try {
    await doLogout();
    recordLog(getDb(), 'INFO', 'whatsapp', 'WhatsApp logout');
    res.json({ ok: true });
  } catch (err) {
    logError(getDb(), 'whatsapp', `Logout error: ${err.message}`);
    res.status(500).json({ error: err.message });
  }
});

// POST /api/whatsapp/notify — send transaction notification
router.post('/whatsapp/notify', requireAuth, requireAdmin, async (req, res) => {
  try {
    const { transaction_id } = req.body;
    const { enabled, autoNotify, ownerNumber } = getWhatsAppSettings();

    if (!enabled || !autoNotify) {
      return res.json({ sent: false, reason: 'disabled' });
    }
    if (!ownerNumber) {
      return res.json({ sent: false, reason: 'missing_owner_number' });
    }
    if (!sock || connectionStatus !== 'ready') {
      return res.json({ sent: false, reason: 'sidecar_not_ready' });
    }

    const db = getDb();
    const trx = db.prepare(
      'SELECT invoice_no, type, total_amount, customer_name, status, created_at FROM transactions WHERE id = ?'
    ).get(transaction_id);

    if (!trx) {
      return res.status(404).json({ sent: false, reason: 'transaction_not_found' });
    }

    const customer = trx.customer_name || '-';
    let message;
    if (trx.type === 'pos') {
      message = `[NOTIFIKASI TRANSAKSI POS]\n\nInvoice: ${trx.invoice_no}\nTipe: Penjualan POS\nNominal: Rp${Math.round(trx.total_amount)}\nPelanggan: ${customer}\nStatus: ${trx.status}\nTanggal: ${trx.created_at}\n\nTercatat otomatis oleh CatatAgen Local`;
    } else if (trx.type === 'brilink') {
      message = `[NOTIFIKASI TRANSAKSI BRILINK]\n\nInvoice: ${trx.invoice_no}\nTipe: Layanan Agen\nNominal: Rp${Math.round(trx.total_amount)}\nPelanggan: ${customer}\nStatus: ${trx.status}\nTanggal: ${trx.created_at}\n\nTercatat otomatis oleh CatatAgen Local`;
    } else {
      message = `[NOTIFIKASI TRANSAKSI]\n\nInvoice: ${trx.invoice_no}\nTipe: ${trx.type}\nNominal: Rp${Math.round(trx.total_amount)}\nPelanggan: ${customer}\nStatus: ${trx.status}\nTanggal: ${trx.created_at}\n\nTercatat otomatis oleh CatatAgen Local`;
    }

    await doSend(ownerNumber, message);
    recordLog(db, 'INFO', 'whatsapp', `WA notif dikirim untuk trx #${transaction_id}`);
    res.json({ sent: true, reason: null, id: null });
  } catch (err) {
    logError(getDb(), 'whatsapp', `Notify error: ${err.message}`);
    res.json({ sent: false, reason: `send_failed: ${err.message}` });
  }
});

export default router;
